Flaky TestNewKubernetesSource/Successful unit test

<!--
Thank you for helping to improve Upjet!

Please be sure to search for open issues before raising a new one. We use issues
for bug reports and feature requests. Please find us at https://slack.crossplane.io
for questions, support, and discussion.
-->

### What happened?
<!--
Please let us know what behaviour you expected and how Upjet diverged from
that behaviour.
-->

Seems that this test is flaky, seen it fail mutiple times over different PRs
```
=== RUN   TestNewKubernetesSource/Successful
    kubernetes_test.go:122: 
        Next(...): -want, +got:
          []migration.UnstructuredWithMetadata{
        - 	{
        - 		Object: unstructured.Unstructured{
        - 			Object: map[string]any{
        - 				"apiVersion": string("ec2.aws.crossplane.io/v1beta1"),
        - 				"kind":       string("VPC"),
        - 				"metadata":   map[string]any{...},
        - 				"spec":       map[string]any{...},
        - 			},
        - 		},
        - 		Metadata: migration.Metadata{Category: s"managed"},
        - 	},
          	{Object: {Object: {"apiVersion": string("azure.crossplane.io/v1beta1"), "kind": string("ResourceGroup"), "metadata": map[string]any{"name": string("example-resources")}, "spec": map[string]any{"forProvider": map[string]any{"location": string("West Europe")}}}}, Metadata: {Category: s"managed"}},
        + 	{
        + 		Object: unstructured.Unstructured{
        + 			Object: map[string]any{
        + 				"apiVersion": string("ec2.aws.crossplane.io/v1beta1"),
        + 				"kind":       string("VPC"),
        + 				"metadata":   map[string]any{...},
        + 				"spec":       map[string]any{...},
        + 			},
        + 		},
        + 		Metadata: migration.Metadata{Category: s"managed"},
        + 	},
          }
--- FAIL: TestNewKubernetesSource (0.00s)
    --- FAIL: TestNewKubernetesSource/Successful (0.00s)
```


### How can we reproduce it?
<!--
Help us to reproduce your bug as succinctly and precisely as possible. Artifacts
such as example manifests or a script that triggers the issue are highly
appreciated!
-->
I assume just running unit tests multiple times should trigger it. I see if fairly often on PRs

--- a/pkg/migration/kubernetes_test.go
+++ b/pkg/migration/kubernetes_test.go
@@ -5,13 +5,17 @@
 package migration
 
 import (
+	"cmp"
+	"slices"
 	"testing"
 
-	"github.com/google/go-cmp/cmp"
+	"github.com/crossplane/crossplane-runtime/pkg/test"
+	gocmp "github.com/google/go-cmp/cmp"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/discovery/cached/memory"
 	"k8s.io/client-go/dynamic/fake"
 	fakeclientset "k8s.io/client-go/kubernetes/fake"
@@ -119,12 +123,34 @@ func TestNewKubernetesSource(t *testing.T) {
 			}
 
 			ks, err := NewKubernetesSource(dynamicClient, memory.NewMemCacheClient(client.Discovery()), WithRegistry(r))
-			if diff := cmp.Diff(tc.want.err, err); diff != "" {
+			if diff := gocmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
 				t.Errorf("\nNext(...): -want, +got:\n%s", diff)
 			}
-			if diff := cmp.Diff(tc.want.ks.items, ks.items); diff != "" {
+			slices.SortFunc(tc.want.ks.items, compareUnstructuredWithMetadata)
+			slices.SortFunc(ks.items, compareUnstructuredWithMetadata)
+			if diff := gocmp.Diff(tc.want.ks.items, ks.items); diff != "" {
 				t.Errorf("\nNext(...): -want, +got:\n%s", diff)
 			}
 		})
 	}
+}
+
+// compareUnstructuredWithMetadata is used for sorting UnstructuredWithMetadata
+// slices based on their namespaced names and GVKs.
+func compareUnstructuredWithMetadata(a, b UnstructuredWithMetadata) int {
+	switch {
+	case a.Object.Object == nil && b.Object.Object == nil:
+		return 0 // not comparable
+	case a.Object.Object == nil && b.Object.Object != nil:
+		return -1
+	case a.Object.Object != nil && b.Object.Object == nil:
+		return 1
+	}
+	fS := func(x *UnstructuredWithMetadata) string {
+		return types.NamespacedName{
+			Namespace: x.Object.GetNamespace(),
+			Name:      x.Object.GetName(),
+		}.String() + ":" + x.Object.GetObjectKind().GroupVersionKind().String()
+	}
+	return cmp.Compare(fS(&a), fS(&b))
 }

